### ISSUE TITLE
Orchestrate runner not resolving reference to a built in state

### ISSUE THREAD

**Issue by llevar** [saltstack/salt#27460]
I have some custom states related to setting up GlusterFS that invoke the builtin glusterfs state. When these were all invoked as part of the top file all was peachy (except the event ordering that I need). I created an orchestrate sls file and now my references to the built in glusterfs state are not being resolved.

This is the orchestrate state file /srv/salt/orch/genotyper.sls

```saltstack
gluster_setup:
  salt.state:
    - tgt: 'roles:glusterfs-server'
    - tgt_type: grain
    - highstate: True
gluster_volume_setup:
  salt.state:
    - tgt: 'roles:glusterfs-master'
    - tgt_type: grain
    - sls: 
      - gluster.master
gluster_volume_mount:
  salt.state:
    - tgt: 'roles:glusterfs-server'
    - tgt_type: grain
    - sls:
      - gluster.mount
```

This is the custom state file /srv/salt/gluster/master.sls
```saltstack
{%- set servers = salt['mine.get']('roles:glusterfs-server', 'network.get_hostname', 'grain_pcre').values() %}
blah:
  cmd.run:
    - name: echo '{{servers}}'

cluster-peers:
  glusterfs.peered:
    - names:
{%- for server in servers %}
      - {{ server }}
{%- endfor %}

cluster-volume:
  glusterfs.created:
    - name: share
    - bricks:
{%- for server in servers %}
      - {{ server }}:/mnt/gluster/brick1
{%- endfor %}
    - stripe: {{ servers|length }}
    - start: True
    - require:
      - glusterfs: cluster-peers
    
```
The command I use is

```
salt-run state.orchestrate orch.genotyper -l debug
```

This is the error that I get:

```yaml
[ERROR   ] Run failed on minions: salt-master
Failures:
    salt-master:
    ----------
              ID: blah
        Function: cmd.run
            Name: echo '['genotyper-1', 'genotyper-0', 'genotyper-2']'
          Result: True
         Comment: Command "echo '['genotyper-1', 'genotyper-0', 'genotyper-2']'" run
         Started: 14:30:30.997670
        Duration: 5.931 ms
         Changes:   
                  ----------
                  pid:
                      12539
                  retcode:
                      0
                  stderr:
                  stdout:
                      [genotyper-1, genotyper-0, genotyper-2]
    ----------
              ID: cluster-peers
        Function: glusterfs.peered
            Name: genotyper-1
          Result: False
         Comment: State 'glusterfs.peered' was not found in SLS 'gluster.master'
                  Reason: 'glusterfs' __virtual__ returned False
         Started: 
        Duration: 
         Changes:   
    ----------
              ID: cluster-peers
        Function: glusterfs.peered
            Name: genotyper-0
          Result: False
         Comment: State 'glusterfs.peered' was not found in SLS 'gluster.master'
                  Reason: 'glusterfs' __virtual__ returned False
         Started: 
        Duration: 
         Changes:   
    ----------
              ID: cluster-peers
        Function: glusterfs.peered
            Name: genotyper-2
          Result: False
         Comment: State 'glusterfs.peered' was not found in SLS 'gluster.master'
                  Reason: 'glusterfs' __virtual__ returned False
         Started: 
        Duration: 
         Changes:   
    ----------
```

Why is it looking for the definition of the glusterfs.peered state inside my custom gluster.master state?

When my top file was like this, there was no problem:

```saltstack
base:
  'G@roles:glusterfs-server':
    - gluster
    - gluster.mount
  'G@roles:glusterfs-master':
    - gluster.master
```

The state definition is definitely present on the host:

```console
[root@salt-master state]# find / -name glusterfs.py
/usr/lib/python2.7/site-packages/salt/modules/glusterfs.py
/usr/lib/python2.7/site-packages/salt/states/glusterfs.py
```

**Comment by jfindlay** [comment]
@llevar, thanks for the report.  What version of salt are you using?

**Comment by llevar** [comment]
```
[root@salt-master state]# salt --versions
Salt Version:
           Salt: 2015.8.0
 
Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.3.1
         Python: 2.7.5 (default, Jun 24 2015, 00:41:19)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 3.2.5
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: 0.5.4
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.1
        timelib: Not Installed
 
System Versions:
           dist: centos 7.1.1503 Core
        machine: x86_64
        release: 3.10.0-229.11.1.el7.x86_64
         system: CentOS Linux 7.1.1503 Core
```

**Comment by jfindlay** [comment]
@llevar, I am unsure why it is referencing `glusterfs.peered`.  I could guess that it is loading the first function in the module it can find.  The error message comes from the [state compiler](https://github.com/saltstack/salt/blob/v2015.8.0/salt/state.py#L858-L875).  Initially the cause seems to be that salt can't load the glusterfs state.  When this happens, it is usually because the corresponding exec or state module doesn't load (`__virtual__()` returns `False`), but that doesn't make sense because you can run the same states using a top file.  Does
```
salt -G 'roles:glusterfs-server' sys.doc gluster.mount
```
return the documentation on that exec module function?

**Comment by llevar** [comment]
That command returns no output. Is this not to be expected? gluster.mount is defined in my state file and it doesn't have any documentation. Also, it's a state not a module as far as I can tell and it looks like sys.doc returns docs for modules. When I run 

```
salt -G 'roles:glusterfs-server' sys.doc glusterfs
```

It returns the docs for the glusterfs module, but not the glusterfs state.

**Comment by jfindlay** [comment]
@llevar, yes, sorry, that is what I meant.  Using `sys.doc` is an easy way to determine if the corresponding execution module was successfully loaded.  Since that is getting loaded, the state module is as well because its `__virtual__` function checks for the presence of the glusterfs exec module, which rules out a platform issue.

**Comment by llevar** [comment]
Ok, this issue was my fault. The minion that was executing glusterfs.peered did not actually have the glusterfs-server package installed. My bad. Although, I think the error message may have been more intuitive.

**Comment by jfindlay** [comment]
@llevar, I'm glad you got it figured out.  Do you mind if we reopen this for producing better error messages in this situation?  Also, it will probably be more reliable to check each glusterfs exec module function against the utilities it uses rather than or in addition to a single `__virtual__` function for the whole module since the module uses utilities from potentially more than one package.

**Comment by DanyC97** [comment]
+1 @jfindlay

**Comment by llevar** [comment]
+1

**Comment by cachedout** [comment]
Closed by #27509
